accepted/2022/ReflectionInvoke.md

```diff
@@ -0,0 +1,354 @@
+        mi2.InvokeDirect(target: default, arg1: tr);
+
+        // Just like these methods can be called today:
+        ChangeIt1(span);
+        ChangeIt2(ref span);
+    }
+
+    public static void ChangeIt1(Span<int> span) { }
+    public static void ChangeIt2(ref Span<int> span) { }
+}
+```
+
+FWIW `__arglist` currently compiles with byref-like types:
+```cs
+Span<byte> s1 = stackalloc byte[1];
+Span<byte> s2 = stackalloc byte[11];
+Span<byte> s3 = stackalloc byte[111];
+CallMe(__arglist(s1, s2, s3));
+
+static unsafe void CallMe(__arglist)
+{
+    // However, when enumerating __arglist here, the runtime throws when accessing a byref-like
+    // type although that limitation could be removed (just a runtime limitation; not compiler)
+```
+
+## Variable-length, safe collections
+(todo; see https://github.com/dotnet/runtime/issues/75349. _I have a local prototype that does a callback that doesn't require any new language or runtime features however it is not a clean programming model. The callback is necessary to interop with hide our internal GC "tracking" feature via "RegisterForGCReporting()"._)
+
+## Exception handling
+Todo; discuss using `BindingFlags.DoNotWrapExceptions` semantics only (no wrapping of exceptions).
+
+# Proposed APIs
+(work in progress; various prototypes exist here)
+## TypedReference
+```diff
+namespace System
+{
+    public ref struct TypedReference
+    {
+        // Equivalent of __makeref (except for a byref-like type since they can't be a generic parameter)
++       public static TypedReference Make<T>(ref T? value);
+        // Helper used for boxed or loosely-typed cases
++       public static TypedReference Make(ref object value, Type type);
+
+        // Equivalent of __refvalue
++       public ref T GetValue<T>();
+        // Used for byref-like types or loosely-typed cases with >4 parameters
++       public readonly unsafe ref byte TargetRef { get; }
+    }
+}
+```
+
+## MethodInfo
+(these may be extension methods instead)
+
+```diff
+namespace System.Reflection
+{
+    public abstract class MethodBase
+    {
+        // Helpers for <= 4 parameters
+
+        // Note that 'default(TypedReference)' can be specified for any "not used" arguments, such as for
+        // static methods ('target' is 'default'), 'void'-returning methods ('result' is 'default') or
+        // for any trailing unused arguments (e.g. 'arg4' can be 'default').
+
+        // We may also want to consider using 'ref byte' instead of or in addition to 'TypedReference'
+        // along with new helper methods like 'public static ref byte GetRawData(object o)'
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference result);
+
++       [System.CLSCompliantAttribute(false)]
++       public virtual void InvokeDirect(
++           TypedReference target,
++           TypedReference arg1,
++           TypedReference result);
```

Comment:
What will be TypedReference value of method return a void?

Reply:
I'm thinking just `default(TypedReference)` would be used. I believe it mentions this just above.